src/editor/Gui.cpp

```diff
@@ -219,7 +228,113 @@ void Gui::pasteLightmap() {
 	dst.nLightmapOffset = src.nLightmapOffset;
 	memcpy(dst.nStyles, src.nStyles, 4);
 
-	app->mapRenderers[app->pickInfo.mapIdx]->reloadLightmaps();
+	app->getSelectedMap()->reloadLightmaps();
+}
+
+
+void ExportModel(Bsp* map, int id)
+{
+	map->update_ent_lump();
+
+	Bsp* tmpMap = new Bsp(map->path);
+	tmpMap->is_model = true;
+
+	BSPMODEL tmpModel = map->models[id];
+
+	Entity* tmpEnt = new Entity(*map->ents[0]);
+
+	vec3 EntOffset = vec3();
+
+	for (int i = 0; i < map->ents.size(); i++)
+	{
+		if (map->ents[i]->getBspModelIdx() == id)
+		{
+			EntOffset = map->ents[i]->getOrigin();
+			break;
+		}
+	}
+
+	vec3 modelOrigin = map->get_model_center(id);
+
+	tmpMap->modelCount = 1;
+	tmpMap->models[0] = tmpModel;
+
+	// Move model to 0 0 0
+	tmpMap->move(-modelOrigin, 0, true);
+
+	for (int i = 1; i < tmpMap->ents.size(); i++)
+	{
+		delete tmpMap->ents[i];
+	}
+
+	tmpMap->ents.clear();
+
+	tmpEnt->setOrAddKeyvalue("origin", vec3(0,0,0).toKeyvalueString());
+	tmpEnt->setOrAddKeyvalue("compiler", g_version_string);
+	tmpEnt->setOrAddKeyvalue("message", "bsp model");
+	tmpMap->ents.push_back(tmpEnt);
+
+	tmpMap->update_ent_lump();
+
+	tmpMap->lumps[LUMP_MODELS] = (BYTE*)tmpMap->models;
+	tmpMap->header.lump[LUMP_MODELS].nLength = sizeof(tmpModel);
+	tmpMap->update_lump_pointers();
+
+	/*for (int i = 0; i < tmpMap->leafCount; i++)
+	{
+		BSPLEAF& tmpLeaf = tmpMap->leaves[i];
+		tmpLeaf.nVisOffset = -1; 
+		for (int n = 0; n < 3; n++)
+		{
+			tmpLeaf.nMins[n] = tmpLeaf.nMins[n] < tmpModel.nMins[n] ? tmpModel.nMins[n] : tmpLeaf.nMins[n];
+			tmpLeaf.nMaxs[n] = tmpLeaf.nMaxs[n] > tmpModel.nMaxs[n] ? tmpModel.nMaxs[n] : tmpLeaf.nMaxs[n];
+		}
+	}*/
+
+	STRUCTCOUNT removed = tmpMap->remove_unused_model_structures();
+	if (!removed.allZero())
+		removed.print_delete_stats(1);
+
+	if (!tmpMap->validate())
+	{
+		int markid = 0;
+		for (int i = 0; i < tmpMap->leafCount; i++)
+		{
+			BSPLEAF& tmpLeaf = tmpMap->leaves[i];
+			tmpLeaf.iFirstMarkSurface = markid;
+			markid += tmpLeaf.nMarkSurfaces;
+		}
+
+		while (tmpMap->models[0].nVisLeafs >= tmpMap->leafCount)
+			tmpMap->create_leaf(-2);
+	}

```

Comment:
@wootguy I don't know why, but some models can't be validated because no leaf count matches.
@wootguy You can tell me how to fix it without these code ?